src/QsCompiler/BondSchemas/CompilerDataStructures.bond

```diff
@@ -0,0 +1,911 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.BondSchemas;
+
+// Data structures from DataTypes.fs
+
+struct Position
+{
+    5: required int32 Line;
+
+    10: required int32 Column;
+}
+
+struct Range
+{
+    5: required Position Start;
+
+    10: required Position End;
+}
+
+// Data structures from SyntaxTokens.fs
+
+enum AccessModifier
+{
+    DefaultAccess,
+    Internal
+}
+
+struct Modifiers
+{
+    5: required AccessModifier Access = DefaultAccess;
+}
+
+enum QsGeneratorDirective
+{
+    SelfInverse,
+    Invert,
+    Distribute,
+    InvalidGenerator
+}
+
+enum QsResult
+{
+    Zero,
+    One
+}
+
+enum QsPauli
+{
+    PauliX,
+    PauliY,
+    PauliZ,
+    PauliI
+}
+
+// N.B. Does not completely represent F#'s QsExpressionKind.
+//      Used by QsExpressionKindComposition to do a complete representation of the original F# type.
+enum QsExpressionKind
+{
+    UnitValue,
+    Identifier,
+    ValueTuple,
+    IntLiteral,
+    BigIntLiteral,
+    DoubleLiteral,
+    BoolLiteral,
+    StringLiteral,
+    ResultLiteral,
+    PauliLiteral,
+    RangeLiteral,
+    NewArray,
+    ValueArray,
+    ArrayItem,
+    NamedItem,
+    NEG,
+    NOT,
+    BNOT,
+    ADD,
+    SUB,
+    MUL,
+    DIV,
+    MOD,
+    POW,
+    EQ,
+    NEQ,
+    LT,
+    LTE,
+    GT,
+    GTE,
+    AND,
+    OR,
+    BOR,
+    BAND,
+    BXOR,
+    LSHIFT,
+    RSHIFT,
+    CONDITIONAL,
+    CopyAndUpdate,
+    UnwrapApplication,
+    AdjointApplication,
+    ControlledApplication,
+    CallLikeExpression,
+    MissingExpr,
+    InvalidExpr
+}
+
+// Used to encapsulate data when QsExpressionKind is Identifier.
+// N.B. Does not have an equivalent F# type.
+struct QsExpressionKindIdentifier<TSymbol, TType>
+{
+    5: required TSymbol Symbol;
+
+    10: nullable<vector<TType>> Types = nothing;
```

Comment:
In this case the distinction between an empty vector and the absence of a vector is needed since the purpose is to represent the `QsExpressionKind` as `Identifier` which has the following definition: `Identifier            of 'Symbol * QsNullable<ImmutableArray<'Type>>`. 

The idea is to revise in more detail the compiler data structures to make them more efficient, but as a first step we want to be able to translate back and forth as accurately as possible because we currently don't know what side effects would arise from slight changes like this one.